### PR TITLE
bpf: fixing dirty cb state from the kernel

### LINKED ISSUE #272: netperf fails to run if container's MTU is -50 Bytes than external interface
Setup:
- External interface with a MTU of 1460
- container's MTU set with 1410 (-50 Bytes)
- cilium running with ipv4 and vxlan mode ON
```
$ ip l 
2: ens4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1460 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether 42:01:ac:10:00:14 brd ff:ff:ff:ff:ff:ff
6: cilium_net@cilium_host: <BROADCAST,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether ce:f9:a6:c3:81:98 brd ff:ff:ff:ff:ff:ff
7: cilium_host@cilium_net: <BROADCAST,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether 8a:c4:b0:1e:67:7b brd ff:ff:ff:ff:ff:ff
8: cilium_vxlan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/ether 06:18:46:0d:cc:03 brd ff:ff:ff:ff:ff:ff
102: lxc92f04@if101: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1410 qdisc noqueue state UP mode DEFAULT group default 
    link/ether 8e:42:fa:41:87:b3 brd ff:ff:ff:ff:ff:ff link-netnsid 0
```

Run `netperf -l 30 -t TCP_STREAM -H SOME_IPV4` the test fails after 4 to 7 seconds.

With 1420 everything runs fine, with IPv6 (in VxLAN mode)+1410 runs fine, this only happens when we aren't fragmenting VxLAN packets (IPv4 in VxLAN mode).

### PR DESCRIPTION
**Pull request by aanm**
Fixes #272
Signed-off-by: André Martins <andre@cilium.io>

### CODE REVIEW

File: bpf/bpf_lb.c [hunk 1]
```diff
@@ -169,7 +169,12 @@ static inline int handle_ipv4(struct __sk_buff *skb)
 __section("from-netdev")
 int from_netdev(struct __sk_buff *skb)
 {
-	int ret;
+	int ret = 0;
+	skb->cb[CB_SRC_LABEL] = ret;
+	skb->cb[CB_IFINDEX] = ret;
+	skb->cb[CB_POLICY] = ret;
+	skb->cb[CB_NAT46_STATE] = ret;
+	skb->cb[CB_CT_STATE] = ret;c
```
Comment (mchalla) [inline bpf/bpf_lb.c]: The c looks like a typo.

File: bpf/bpf_lxc.c [hunk 2]
```diff
@@ -562,7 +562,12 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_ARP) int tail_handle_arp(struct __s
 __section("from-container")
 int handle_ingress(struct __sk_buff *skb)
 {
-	int ret;
+	int ret = 0;
+	skb->cb[CB_SRC_LABEL] = ret;
+	skb->cb[CB_IFINDEX] = ret;
+	skb->cb[CB_POLICY] = ret;
+	skb->cb[CB_NAT46_STATE] = ret;
+	skb->cb[CB_CT_STATE] = ret;c
```
Comment (mchalla) [inline bpf/bpf_lxc.c]: May be we should add an inline function or macro to common.h instead of repeating this code in 4 places.

Comment (borkmann) [reply bpf/bpf_lxc.c]: Good point. We should perhaps write it without the enum, f.e.: __u32 zero = 0; skb->cb[0] = zero; skb->cb[1] = zero; ... skb->cb[4] = zero, as that is ABI anyway and independent of CB_*.

**Review by mchalla — APPROVED** [review]

File: bpf/bpf_lb.c [hunk 3]
```diff
@@ -169,6 +169,8 @@ static inline int handle_ipv4(struct __sk_buff *skb)
 __section("from-netdev")
 int from_netdev(struct __sk_buff *skb)
 {
+	bpf_clear_cb(skb);
```
Comment (mchalla) [inline bpf/bpf_lb.c]: The int ret should be on top. I approved it. You can change it.

Comment (aanm) [reply bpf/bpf_lb.c]: oh I thought this should be the first thing ever. OK, I'll change it ;)

**Comment by borkmann** [comment]
Looks good now, thanks!

Signed-off-by: Daniel Borkmann <daniel@cilium.io>

**Review by borkmann — APPROVED** [review]

**Pull request merged**